src/extensions/default/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/PackageUpdaterStep.cs

```diff
@@ -150,6 +155,46 @@ void AddSubsteps<T>(IEnumerable<Operation<T>> items, string messageFormat, Func<
                 return new UpgradeStepInitializeResult(UpgradeStepStatus.Incomplete, $"{_analyzer.Name} has identified some recommended changes", Risk: _risk);
             }
 
+            // For server-only projects, do not need to add System.ServiceModel packages.
+            private static IEnumerable<Operation<NuGetReference>> UpdatePackageAddition(IDependencyCollection<NuGetReference> packages, IProject project)
+            {
+                var files = project.FindFiles(".cs", ProjectItemType.Compile);
+                var containsService = false;
+                if (files.Any())
+                {
+                    foreach (var f in files)
+                    {
+                        var root = CSharpSyntaxTree.ParseText(f).GetRoot();
+                        if (ContainsIdentifier(root, "ChannelFactory") || ContainsIdentifier(root, "ClientBase"))
+                        {
+                            return packages.Additions;
+                        }
+
+                        if (ContainsIdentifier(root, "ServiceHost"))
+                        {
+                            containsService = true;
+                        }
+                    }
+
+                    if (containsService)
+                    {
+                        return from p in packages.Additions
+                               where !p.Item.Name.Contains("System.ServiceModel")
+                               select p;
+                    }
+                }
+
+                return packages.Additions;
+            }
+
+            // Checks if the root has descendant nodes that contains id
+            private static bool ContainsIdentifier(SyntaxNode root, string id)
+            {
+                return (from n in root.DescendantNodes().OfType<IdentifierNameSyntax>()
+                        where n.Identifier.ValueText.Contains(id)
+                        select n).Any();
```

Comment:
I think you want to use StartsWith for the same performance reasons as mentioned above. As class names are case sensitive, you should pass StringComparison.Ordinal.

Reply:
Additionally. Instead of `where` then `Any` you can use lambda syntax an change this whole statement to read:
```csharp
return root.DescendantNodes().OfType<IdentifierNameSyntax>().Any(n => n.Identifier.ValueText.StartsWith(id, StringComparison.Ordinal);
```

Reply:
That's a good point. You still need to use StringComparison.Ordinal otherwise it needs to do more work when doing comparisons as there are some characters considered equivalent in some languages such as accented characters. Ordinal simply means exact same character ignoring any nuances of the current cultures language.